include/unifex/let_with_stop_source.hpp

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/config.hpp>
+#include <unifex/just.hpp>
+#include <unifex/let.hpp>
+#include <unifex/receiver_concepts.hpp>
+#include <unifex/sender_concepts.hpp>
+#include <unifex/inplace_stop_token.hpp>
+#include <unifex/execution_policy.hpp>
+#include <unifex/get_execution_policy.hpp>
+#include <unifex/type_list.hpp>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+
+namespace _let_with_stop_source {
+
+template<typename Operation, typename Receiver>
+struct _stop_source_receiver {
+    class type;
+};
+
+template<typename InnerOp, typename Receiver>
+struct _stop_source_operation {
+  struct type;
+};
+
+template <typename InnerOp, typename Receiver>
+using operation =  typename _stop_source_operation<InnerOp, remove_cvref_t<Receiver>>::type;
+
+
+template<typename Operation, typename Receiver>
+using stop_source_receiver = typename _stop_source_receiver<Operation, Receiver>::type;
+
+template<typename Operation, typename Receiver>
+class _stop_source_receiver<Operation, Receiver>::type {
+public:
+    explicit type(Operation& op, Receiver&& r) :
+        op_(op), receiver_{std::forward<Receiver>(r)}
+    {}
+
+    template<typename... Values>
+    void set_next(Values&&... values) &
+        noexcept(is_nothrow_next_receiver_v<Receiver, Values...,  unifex::inplace_stop_source&>) {
+
+        // Forward, injecting stop_source
+        unifex::set_next(receiver_, std::forward<Values>(values)..., op_.stop_source_);
+    }
```

Comment:
Do we need to make this algorithm a many-sender to inject the stop-source via set_next()?

Can we instead just rely on the injection of the stop-source via the SuccessorFactory and make this algorithm a single-sender-only algorithm?

Usually, any algorithm that is injecting cancellation of a bulk operation will also control the reduce operation and so will be able to apply this algorithm outside, eg. the `bulk_join()`.

Reply:
No we don't. I don't think that code is used, I must have not deleted it when cleaning up.